No data view

e18r opened the pull request:
Fixes autarklabs/open-enterprise#1713
![Screenshot from 2020-01-08 10-26-16](https://user-images.githubusercontent.com/16065447/71990437-5d4e7380-3201-11ea-9dc0-13ad7cfbf04d.png)

PeterMPhillips commented on app/src/App.js:
```diff
@@ -58,6 +61,53 @@ function App() {
+      <SidePanelContainer>
+        <PanelContent
+          ipfsAddr={
+            entries !== undefined &&
```
```suggestion
            entries.length !== 0 &&
```

PeterMPhillips commented on app/src/App.js:
```diff
@@ -58,6 +61,53 @@ function App() {
+            entries[selectedWidget].addr
+          }
+          content={
+            entries !== undefined &&
```
```suggestion
            entries.length !== 0 &&
```

PeterMPhillips commented on app/src/App.js:
```diff
@@ -58,6 +61,53 @@ function App() {
+    </SidePanel>
+  )
+
+  if (!entries) {
```
```suggestion
  if (entries.length === 0) {
```

PeterMPhillips reviewed (changes requested):
Looks good! The only issue I have is the assumption that `entries = undefined` we should set the default entries value to an empty array on line 35: `const { entries = [] } = appState` and then make the following changes:

chadoh commented on app/src/App.js:
```diff
@@ -58,6 +61,53 @@ function App() {
   //   setEditMode(!editMode)
   // }
 
+  const SideContent = () => (
```
I've seen lots of problems caused by defining components within other components.

Every time React renders `App`, it's going to redefine `SideContent`, and that means that all elements rendered as part of SideContent will be removed from the DOM and re-added. This is incredibly inefficient and can cause big UX problems – in Projects, we had a spot where an input was removed from the DOM and re-added after every keystroke! For the user, this meant that the input lost focus every time they typed a character in it! Incredibly frustrating.

I get that it's tempting to do this. We don't have to pass in any props! But the tradeoffs are almost never worth it.

chadoh commented on app/src/App.js:
```diff
@@ -159,6 +172,13 @@ const BaseLayout = styled.div`
 
+const EmptyLayout = styled.div`
+  display: flex;
+  height: 95vh;
```
Where does 95 come from? What takes up the other 5?

chadoh reviewed (changes requested):
A couple things to clean up here

e18r committed:
fix: SidePanel as a separate component & other fixes

* `SidePanel` as a separate component
* `entries` initialized as an empty array if undefined
* Empty card with description text
* `SideContent` prop validation fix

e18r committed:
fix: remove scrollbar with 100% height

chadoh commented on app/src/App.js:
```diff
@@ -58,6 +64,44 @@ function App() {
+                  color={`${theme.surfaceContentSecondary}`}
+                >
+                  Present important information to current and prospective
+                  members of your organization.
```
I vote we leave off the period at the end of this sentence – I don't think we include it elsewhere, and it has [stern connotations with a growing number of English speakers](https://www.npr.org/transcripts/747020219). @javieralaves: thoughts?

chadoh replied:
This is still a relevant question, even though GitHub marks it as outdated. The line moved, but the period is still there. Should we remove it, @javieralaves?

chadoh replied:
I've removed the period now, so I'm happy with this PR

chadoh commented on app/src/App.js:
```diff
@@ -58,6 +64,44 @@ function App() {
+          newWidget={newWidget}
+          updateWidget={updateWidget}
+        />
+      </Main>
```
The `SideContent` and `Main` are basically the same between this early return and the main return, but that's really hard to see now. If we change `SideContent` in one spot, we will also need to change it in the other, and it will be easy to forget that and end up with bugs.

chadoh commented on app/src/App.js:
```diff
@@ -58,6 +64,44 @@ function App() {
+  }
+
   const widgetList =
     entries &&
```
We know that `entries` is always defined now. We can remove this first check.

Also, why define `widgetList` up here, rather than inline? It's only used in one place.

chadoh committed:
fix: extracted Routes component with early return

Rather than duplicating SideContent in each return from the main App.js
function, this keeps all logic for rendering SideContent in one spot and
instead moves the logic for what else to render into the Routes
component. This component is responsible for checking if entries is
empty and returning early with the empty card if so.

Other cleanup:

* remove commented-out code
* use `function` fore all components
* put `SideContent` before `App` – not sure about this one
* re-retrieve `entries` from `appState` rather than passing as prop

chadoh reviewed (changes requested):
I pushed my own commit with some cleanup I wanted to see – see obsolete comments below.

I noticed one other thing I'd like us to address now if we can. Does it seem like the side panel is opening too fast? The animation of the background fading happens much more slowly than the side panel opening. See GIF:

![side panel opening too fast](https://user-images.githubusercontent.com/221614/72002728-f6878500-3215-11ea-92ba-90fb782d1bdf.gif)

e18r commented:
@chadoh I remember this happened already in the past at some point where I think we introduced a Wrapper component around some shared logic. I'm gonna check it out and let you know.

e18r commented:
I sort of tracked down the origin of the side panel appearing without animation. It's something so unrelated to this PR, that I think it's worth creating [a separate issue](https://github.com/AutarkLabs/open-enterprise/issues/1854).

ottodevs commented:
This PR was merged into #47
